File: src/java/com/kitware/pulse/engine/PulseScenarioExec.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.engine;

import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.util.JsonFormat;
import com.kitware.pulse.cdm.bind.Scenario.eSerializationFormat;
import com.kitware.pulse.cdm.scenario.SEScenarioExec;
import com.kitware.pulse.engine.bind.Enums.eModelType;
import com.kitware.pulse.engine.bind.Scenario.ScenarioExecData;
import com.kitware.pulse.utilities.Log;
import com.kitware.pulse.utilities.LogListener;

/**
 * This is a class that will run a scenario through Pulse
 * and create it's results file. You could easily write your
 * own exec class that does what you want in between actions
 * but this may work for most of your needs.
 * @author abray
 */
public class PulseScenarioExec extends SEScenarioExec
{
  protected static class MyListener extends LogListener
  {
    public String origin = "";
    protected MyListener()
    {
      super();
      listen(false);
    }    
    @Override public void handleDebug(String msg) { Log.debug(origin+msg); }
    @Override public void handleInfo(String msg)  { Log.info(origin+msg); }
    @Override public void handleWarn(String msg)  { Log.warn(origin+msg); }
    @Override public void handleError(String msg) { Log.error(origin+msg); }
    @Override public void handleFatal(String msg) { Log.fatal(origin+msg); }
  }
  
  protected MyListener logListener = new MyListener();
  protected eModelType modelType;

  public PulseScenarioExec()
  {
    clear();
  }

  public void clear()
  {
    super.clear();
    modelType = eModelType.HumanAdultWholeBody;
  }
  
  public void copy(PulseScenarioExec src)
  {
    PulseScenarioExec.load(PulseScenarioExec.unload(src), this);
  }
  
  public static void load(ScenarioExecData src, PulseScenarioExec dst)
  {
    SEScenarioExec.load(src.getScenarioExec(), dst);
    dst.modelType = src.getModelType();
  }
  public static ScenarioExecData unload(PulseScenarioExec src)
  {
    ScenarioExecData.Builder dst = ScenarioExecData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(PulseScenarioExec src, ScenarioExecData.Builder dst)
  {
    SEScenarioExec.unload(src, dst.getScenarioExecBuilder());
    dst.setModelType(src.modelType);
  }
  
  public String toJSON() 
  {
    String json;
    try
    {
      json = JsonFormat.printer().print(PulseScenarioExec.unload(this));
    }
    catch (InvalidProtocolBufferException e)
    {
      Log.error("Unable to generate json from PulseScenarioExec");
      json = "";
    }
    return json;
  }
  
  public eModelType getModelType() { return modelType; }
  public void setModelType(eModelType mt) { modelType=mt; }
  
  public void setLogPrepend(String lp) { logListener.origin = lp.isEmpty() ? "" : lp+" "; }
  
  public boolean execute()
  {
    return nativeExecuteScenario(toJSON(), eSerializationFormat.JSON_VALUE);
  }
  protected native boolean nativeExecuteScenario(String sceOpts, int format);
  
  ///////////////////////
  // LISTENER/HANDLERS //
  ///////////////////////
  
  protected void handleDebug(String msg)
  {
    if(this.logListener!=null)
    this.logListener.debug(msg);
  }
  protected void handleInfo(String msg)
  {
    if(this.logListener!=null)
    this.logListener.info(msg);
  }
  protected void handleWarning(String msg)
  {
    if(this.logListener!=null)
    this.logListener.warn(msg);
  }
  protected void handleError(String msg)
  {
    if(this.logListener!=null)
    this.logListener.error(msg);
  }
  protected void handleFatal(String msg)
  {
    if(this.logListener!=null)
    this.logListener.fatal(msg);
  }
}

File: src/python/pulse/engine/io/PulseScenarioExec.py
# Distributed under the Apache License, Version 2.0.
# See accompanying NOTICE file for details.

from pulse.engine.PulseScenarioExec import PulseScenarioExec

from pulse.cdm.engine import eSerializationFormat
from pulse.cdm.io.scenario import serialize_scenario_exec_to_bind
from pulse.engine.bind.Scenario_pb2 import ScenarioExecData
from google.protobuf import json_format

def serialize_pulse_scenario_exec_to_string(src: PulseScenarioExec, fmt: eSerializationFormat) -> str:
    dst = ScenarioExecData()
    serialize_pulse_scenario_exec_to_bind(src, dst)
    return json_format.MessageToJson(dst, True, True)

def serialize_pulse_scenario_exec_to_file(src: PulseScenarioExec, filename: str) -> None:
    string = serialize_pulse_scenario_exec_to_string(src, eSerializationFormat.JSON)
    file = open(filename, "w")
    n = file.write(string)
    file.close()

def serialize_pulse_scenario_exec_from_string(string: str, dst: PulseScenarioExec, fmt: eSerializationFormat) -> None:
    src = ScenarioExecData()
    json_format.parse(string, src)
    serialize_pulse_scenario_exec_from_bind(src, dst)

def serialize_pulse_scenario_exec_from_file(filename: str, dst: PulseScenarioExec) -> None:
    with open(filename) as f:
        string = f.read()
    serialize_pulse_scenario_exec_from_string(string, dst, eSerializationFormat.JSON)

def serialize_pulse_scenario_exec_to_bind(src: PulseScenarioExec, dst: ScenarioExecData) -> None:
    serialize_scenario_exec_to_bind(src, dst.ScenarioExec)
    dst.ModelType = src.get_model_type().value

def serialize_pulse_scenario_exec_from_bind(src: ScenarioExecData, dst: PulseScenarioExec):
    raise Exception("serialize_pulse_scenario_exec_from_bind not implemented")
